lib/Dialect/Ibis/Transforms/IbisContainersToHW.cpp

```diff
@@ -0,0 +1,306 @@
+  llvm::DenseMap<StringAttr, InputPortOp> opInputs;
+
+  // A mapping between the port name and the port op within the container.
+  llvm::DenseMap<StringAttr, OutputPortOp> opOutputs;
+
+  ContainerPortInfo() = default;
+  ContainerPortInfo(ContainerOp container) {
+    SmallVector<hw::PortInfo, 4> inputs, outputs;
+
+    // Gather in and output port ops.
+    for (auto input : container.getBodyBlock()->getOps<InputPortOp>()) {
+      opInputs[input.getSymNameAttr()] = input;
+
+      hw::PortInfo portInfo;
+      portInfo.name = input.getSymNameAttr();
+      portInfo.type = cast<PortOpInterface>(input.getOperation()).getPortType();
+      portInfo.dir = hw::ModulePort::Direction::Input;
+      inputs.push_back(portInfo);
+    }
+
+    for (auto output : container.getBodyBlock()->getOps<OutputPortOp>()) {
+      opOutputs[output.getSymNameAttr()] = output;
+
+      hw::PortInfo portInfo;
+      portInfo.name = output.getSymNameAttr();
+      portInfo.type =
+          cast<PortOpInterface>(output.getOperation()).getPortType();
+      portInfo.dir = hw::ModulePort::Direction::Output;
+      outputs.push_back(portInfo);
+    }
+
+    // Sort the in- and output port lists by name to ensure deterministic
+    // ordering.
+    auto cmp = [](const hw::PortInfo *lhs, const hw::PortInfo *rhs) {
+      return lhs->name.getValue().compare(rhs->name.getValue());
+    };
+    llvm::array_pod_sort(inputs.begin(), inputs.end(), cmp);
+    llvm::array_pod_sort(outputs.begin(), outputs.end(), cmp);
+    hwPorts = std::make_unique<hw::ModulePortInfo>(inputs, outputs);
+  }
+};
+
+using ContainerPortInfoMap = llvm::DenseMap<StringAttr, ContainerPortInfo>;
+
+struct ContainerOpConversionPattern : public OpConversionPattern<ContainerOp> {
+  ContainerOpConversionPattern(MLIRContext *ctx,
+                               ContainerPortInfoMap &portOrder)
+      : OpConversionPattern<ContainerOp>(ctx), portOrder(portOrder) {}
+
+  LogicalResult
+  matchAndRewrite(ContainerOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    rewriter.setInsertionPoint(op);
+
+    const ContainerPortInfo &cpi = portOrder.at(op.getSymNameAttr());
+    auto hwMod = rewriter.create<hw::HWModuleOp>(
+        op.getLoc(), op.getSymNameAttr(), *cpi.hwPorts);
+
+    hw::OutputOp outputOp =
+        cast<hw::OutputOp>(hwMod.getBodyBlock()->getTerminator());
+
+    // Replace all of the reads of the inputs to use the input block arguments.
+    for (auto [idx, input] : llvm::enumerate(cpi.hwPorts->getInputs())) {
+      Value barg = hwMod.getBodyBlock()->getArgument(idx);
+      InputPortOp inputPort = cpi.opInputs.at(input.name);
+      // Replace all reads of the input port with the input block argument.
+      for (auto *user : inputPort.getOperation()->getUsers()) {
+        auto reader = dyn_cast<PortReadOp>(user);
+        if (!reader)
+          return rewriter.notifyMatchFailure(
```

Comment:
Should this just be an error instead of a match failure? Nothing can match on it since it's invalid code, yes? It should probably be part of a verifier and this should just be an assert.

Reply:
> It should probably be part of a verifier

I've debated this with myself. We can have a local verification of multiple writers (i.e. in the same scope) but that still doesn't prevent a design from being legal (given our spooky-writing-at-a-distance capability) so i'd just be a partial validity check. 
Changed to emit an error.